torchvision/csrc/io/image/cuda/decode_jpegs_cuda.cpp

```diff
@@ -0,0 +1,582 @@
+
+  std::vector<torch::Tensor> output_tensors{encoded_images.size()};
+  std::vector<nvjpegImage_t> decoded_images{encoded_images.size()};
+
+  for (std::vector<at::Tensor>::size_type i = 0; i < encoded_images.size();
+       i++) {
+    // extract bitstream meta data to figure out the number of channels, height,
+    // width for every image
+    status = nvjpegGetImageInfo(
+        nvjpeg_handle,
+        (unsigned char*)encoded_images[i].data_ptr(),
+        encoded_images[i].numel(),
+        &channels[i],
+        &subsampling,
+        width,
+        height);
+    TORCH_CHECK(
+        status == NVJPEG_STATUS_SUCCESS, "Failed to get image info: ", status);
+
+    TORCH_CHECK(
+        subsampling != NVJPEG_CSS_UNKNOWN, "Unknown chroma subsampling");
+
+    // output channels may be different from the actual number of channels in
+    // the image, e.g. we decode a grayscale image as RGB and slice off the
+    // extra channels later
+    int output_channels = 3;
+    if (output_format == NVJPEG_OUTPUT_RGB ||
+        output_format == NVJPEG_OUTPUT_UNCHANGED) {
+      output_channels = 3;
+    } else if (output_format == NVJPEG_OUTPUT_Y) {
+      output_channels = 1;
+    }
+
+    // reserve output buffer
+    auto output_tensor = torch::empty(
+        {int64_t(output_channels), int64_t(height[0]), int64_t(width[0])},
+        torch::dtype(torch::kU8).device(target_device));
+    output_tensors[i] = output_tensor;
+
+    // fill nvjpegImage_t struct
+    for (int c = 0; c < output_channels; c++) {
+      decoded_images[i].channel[c] = output_tensor[c].data_ptr<uint8_t>();
+      decoded_images[i].pitch[c] = width[0];
+    }
+    for (int c = output_channels; c < NVJPEG_MAX_COMPONENT; c++) {
+      decoded_images[i].channel[c] = NULL;
+      decoded_images[i].pitch[c] = 0;
+    }
+  }
+  return {decoded_images, output_tensors, channels};
+}
+
+std::vector<torch::Tensor> CUDAJpegDecoder::decode_images(
+    const std::vector<torch::Tensor>& encoded_images,
+    const nvjpegOutputFormat_t& output_format) {
+  /*
+    This function decodes a batch of jpeg bitstreams.
+    We scan all encoded bitstreams and sort them into two groups:
+    1. Baseline JPEGs: Can be decoded with hardware support on A100+ GPUs.
+    2. Other JPEGs (e.g. progressive JPEGs): Need to be decoded in software.
+
+    Args:
+    - encoded_images (std::vector<torch::Tensor>): a vector of tensors
+    containing the jpeg bitstreams to be decoded
+    - output_format (nvjpegOutputFormat_t): NVJPEG_OUTPUT_RGB, NVJPEG_OUTPUT_Y
+    or NVJPEG_OUTPUT_
+    - device (torch::Device): The desired CUDA device for the returned Tensors
+
+    Returns:
+    - output_tensors (std::vector<torch::Tensor>): a vector of Tensors
+    containing the decoded images
+  */
+
+  auto [decoded_imgs_buf, output_tensors, channels] =
+      prepare_buffers(encoded_images, output_format);
+
+  nvjpegStatus_t status;
+  cudaError_t cudaStatus;
+
+  cudaStatus = cudaStreamSynchronize(stream);
```

Comment:
Why is this needed?

Reply:
That's how they do it here: https://github.com/NVIDIA/CUDALibrarySamples/blob/f17940ac4e705bf47a8c39f5365925c1665f6c98/nvJPEG/nvJPEG-Decoder/nvjpegDecoder.cpp#L36
After buffers are allocated they synchronize before starting the decoding